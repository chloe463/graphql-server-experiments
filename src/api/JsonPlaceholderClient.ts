import axios, { AxiosInstance } from "axios";
import faker from "faker";
import { NexusGenObjects } from "../graphql/generated/typings";

type Post = NexusGenObjects["Post"];
type Comment = NexusGenObjects["Comment"];

const BASE_URL = "https://jsonplaceholder.typicode.com/";

type FetchPostsParams = {
  start?: string;
  limit?: number;
  page?: number;
};

type FetchPostsResponse = {
  posts: Post[];
  totalCount: number;
};

type FetchCommentsResponse = {
  comments: Comment[];
};

export class JsonPlaceholderClient {
  private client: AxiosInstance;
  constructor() {
    this.client = axios.create({
      baseURL: BASE_URL,
    });
  };

  fetchPosts = async (args: FetchPostsParams): Promise<FetchPostsResponse> => {
    const params = {};
    Object.keys(args).forEach((key) => {
      params[`_${key}`] = args[key];
    });
    const res = await this.client.get<Post[]>("/posts", {
      params: {
        ...params,
      }
    });
    return { posts: res.data, totalCount: res.headers["x-total-count"] };
  };

  makeDummyPosts = (args: FetchPostsParams): FetchPostsResponse => {
    const posts = Array.from({ length: args.limit || 10 }, (_, i) => i + 1).map((i) => {
      const id = parseInt(args.start || "0", 10) + i;
      return {
        id,
        title: faker.lorem.word(10),
        body: faker.lorem.words(100),
        userId: Math.ceil(id / 10),
      };
    });
    return { posts, totalCount: 100 };
  }

  fetchComments = async (postId: number): Promise<FetchCommentsResponse> => {
    const res = await this.client.get<Comment[]>(`/posts/${postId}/comments`);
    return { comments: res.data };
  }

  makeDummyComments = (postId: number): FetchCommentsResponse => {
    const comments: Comment[] = Array.from({ length: 5 }, (_, i) => i + 1).map((i) => {
      const id = (postId - 1) + i;
      return {
        id,
        body: faker.lorem.words(100),
        email: `${faker.name.firstName()}@example.com`,
        name: faker.name.findName(),
        postId,
      }
    });

    return { comments };
  }
}
