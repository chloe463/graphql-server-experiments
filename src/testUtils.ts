import { ApolloServer } from "apollo-server";
import { Context } from "./context";
import { schema } from "./graphql/schema";

type DummyContext = {
  [key in keyof Context]: jest.Mock;
};

export const constructTestServer = (context: DummyContext) => {
  const server = new ApolloServer({
    schema,
    context,
  });

  return server;
};
