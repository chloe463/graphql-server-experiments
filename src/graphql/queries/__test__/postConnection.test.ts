import { gql } from "apollo-server";
import { createTestClient } from "apollo-server-testing";
import { JsonPlaceholderClient } from "../../../api/JsonPlaceholderClient";
import { constructTestServer } from "../../../testUtils";

const jsonPlaceholderClientMock = (JsonPlaceholderClient as any) as jest.Mock<JsonPlaceholderClient>;

const GET_POST_CONNECTION = gql`
  query GetPostConnectionQuery($first: Int, $after: String) {
    postConnection(first: $first, after: $after) {
      pageInfo {
        hasNextPage
        hasPreviousPage
        startCursor
        endCursor
      }
      edges {
        node {
          id
          userId
          title
          body
        }
        cursor
      }
    }
  }
`;

describe("[Query] postConnection", () => {
  it("returns post list", async () => {
    (jsonPlaceholderClientMock as any).fetchPosts = jest.fn().mockReturnValue([
      {
        userId: 1,
        id: 1,
        title: "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
        body: "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto"
      },
      {
        userId: 1,
        id: 2,
        title: "qui est esse",
        body: "est rerum tempore vitae sequi sint nihil reprehenderit dolor beatae ea dolores neque fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis qui aperiam non debitis possimus qui neque nisi nulla"
      },
      {
        userId: 1,
        id: 3,
        title: "ea molestias quasi exercitationem repellat qui ipsa sit aut",
        body: "et iusto sed quo iure voluptatem occaecati omnis eligendi aut ad voluptatem doloribus vel accusantium quis pariatur molestiae porro eius odio et labore et velit aut"
      },
    ]);
    const server = constructTestServer({
      jsonPlaceholderClient: jsonPlaceholderClientMock,
    });

    const { query } = createTestClient(server);
    const res = await query({ query: GET_POST_CONNECTION, variables : { first: 3, after: "0" }});
    console.log({ res });
    expect(res).toMatchSnapshot();
  });
});
