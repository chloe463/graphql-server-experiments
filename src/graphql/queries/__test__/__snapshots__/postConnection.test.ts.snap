// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Query] postConnection returns post list 1`] = `
Object {
  "data": Object {
    "postConnection": Object {
      "edges": Array [
        Object {
          "cursor": "1",
          "node": Object {
            "body": "quia et suscipit suscipit recusandae consequuntur expedita et cum reprehenderit molestiae ut ut quas totam nostrum rerum est autem sunt rem eveniet architecto",
            "id": 1,
            "title": "sunt aut facere repellat provident occaecati excepturi optio reprehenderit",
            "userId": 1,
          },
        },
        Object {
          "cursor": "2",
          "node": Object {
            "body": "est rerum tempore vitae sequi sint nihil reprehenderit dolor beatae ea dolores neque fugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis qui aperiam non debitis possimus qui neque nisi nulla",
            "id": 2,
            "title": "qui est esse",
            "userId": 1,
          },
        },
        Object {
          "cursor": "3",
          "node": Object {
            "body": "et iusto sed quo iure voluptatem occaecati omnis eligendi aut ad voluptatem doloribus vel accusantium quis pariatur molestiae porro eius odio et labore et velit aut",
            "id": 3,
            "title": "ea molestias quasi exercitationem repellat qui ipsa sit aut",
            "userId": 1,
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": null,
        "hasNextPage": true,
        "hasPreviousPage": false,
        "startCursor": null,
      },
      "totalCount": 3,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
